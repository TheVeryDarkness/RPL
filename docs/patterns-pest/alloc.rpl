pattern alloc

patt {
    p[$T: type] = 
        alloc_cast_write[$T = $T] 
        - alloc_check_cast_write[$T = $T] 
        - alloc_cast_check_write[$T = $T] 
        - alloc_cast_check_as_write[$T = $T]
}

util {
    alloc_cast_write[$T: type] = {
        fn _(..) -> _ {
            'alloc:
            let $ptr_1: *mut u8 = alloc::alloc::__rust_alloc(_, _); // _3
            let $ptr_2: *mut $T = move $ptr_1 as *mut $T (PtrToPtr); // _2
            'write:
            (*$ptr_2) = _;
        }
    }
    alloc_check_cast_write[$T: type] = {
        fn _(..) -> _ {
            'alloc:
            let $ptr_1: *mut u8 = alloc::alloc::__rust_alloc(_, _); // _2
            let $const_ptr_1: *const u8 = copy $ptr_1 as *const u8 (PtrToPtr); // _19
            let $addr_1: usize = copy $const_ptr_1 as usize (Transmute); // _20
            // It's weird that `$ptr_2` can only be declared before `switchInt`
            // switchInt(move $addr_1) {
            //     0_usize => {}
            //     _ => {}
            // }
            let $ptr_2: *mut $T = copy $ptr_1 as *mut $T (PtrToPtr); // _4
            'write:
            (*$ptr_2) = _;
        }
    }
    alloc_cast_check_write[$T: type] = {
        fn _(..) -> _ {
            'alloc:
            let $ptr_1: *mut u8 = alloc::alloc::__rust_alloc(_, _); // _3
            let $ptr_2: *mut $T = move $ptr_1 as *mut $T (PtrToPtr); // _2
            let $const_ptr_1: *const u8 = copy $ptr_2 as *const u8 (PtrToPtr); // _19
            let $addr_1: usize = copy $const_ptr_1 as usize (Transmute); // _20
            // switchInt(move $addr_1) {
            //     0_usize => {}
            //     _ => {}
            // }
            'write:
            (*$ptr_2) = _;
        }
    }
    alloc_cast_check_as_write[$T: type] = {
        fn _(..) -> _ {
            'alloc:
            let $ptr_1: *mut u8 = alloc::alloc::__rust_alloc(_, _); // _3
            let $ptr_2: *mut $T = move $ptr_1 as *mut $T (PtrToPtr); // _2
            let $addr_1: usize = copy $ptr_2 as usize (PointerExposeProvenance); // _6
            // switchInt(move $addr_1) {
            //     0_usize => {}
            //     _ => {}
            // }
            'write:
            (*$ptr_2) = _;
        }
    }
}

diag {
    p = {
        primary(write) = "it is an undefined behavior to dereference a null pointer, and `std::alloc::alloc` may return a null pointer",
        label(alloc)   = "pointer created here",
        label(write)   = "pointer used here",
        note           = "See https://doc.rust-lang.org/std/alloc/fn.alloc.html and https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#tymethod.alloc",
        // help           = "check whether it's null before dereferencing",
        name           = "unchecked_allocated_pointer",
        level          = "warn",
    }
}
