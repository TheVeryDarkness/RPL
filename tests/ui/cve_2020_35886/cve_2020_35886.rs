//@ revisions: inline regular
//@[inline] compile-flags: -Z inline-mir=true
//@[regular] compile-flags: -Z inline-mir=false
//@[regular] check-pass
// FIXME: write a non-inline pattern
use std::alloc::{Layout, alloc, alloc_zeroed, dealloc};
use std::ops::{Index, IndexMut, Range};

pub struct Array<T> {
    size: usize,
    ptr: *mut T,
}

impl<T> Array<T> {
    /// Convert to slice
    pub fn to_slice<'a>(&'a self) -> &'a [T] {
        unsafe { std::slice::from_raw_parts(self.ptr as *const T, self.size) }
    }

    /// Convert to mutable slice
    pub fn to_slice_mut<'a>(&'a mut self) -> &'a mut [T] {
        unsafe { std::slice::from_raw_parts_mut(self.ptr, self.size) }
    }

    /// The length of the array (number of elements T)
    pub fn len(&self) -> usize {
        self.size
    }
}

impl<T> Index<usize> for Array<T> {
    type Output = T;

    // #[rpl::dump_mir(dump_cfg, dump_ddg)]
    fn index<'a>(&'a self, idx: usize) -> &'a Self::Output {
        unsafe { self.ptr.wrapping_offset(idx as isize).as_ref() }.unwrap()
        //~[inline]^ERROR: it is an undefined behavior to offset a pointer using an unchecked integer
    }
}

impl<T> IndexMut<usize> for Array<T> {
    fn index_mut<'a>(&'a mut self, idx: usize) -> &'a mut Self::Output {
        unsafe { self.ptr.wrapping_offset(idx as isize).as_mut() }.unwrap()
        //~[inline]^ERROR: it is an undefined behavior to offset a pointer using an unchecked integer
    }
}

impl<T> Index<Range<usize>> for Array<T> {
    type Output = [T];

    fn index<'a>(&'a self, idx: Range<usize>) -> &'a Self::Output {
        &self.to_slice()[idx]
    }
}

impl<T> IndexMut<Range<usize>> for Array<T> {
    fn index_mut<'a>(&'a mut self, idx: Range<usize>) -> &'a mut Self::Output {
        &mut self.to_slice_mut()[idx]
    }
}

impl<T> Drop for Array<T> {
    fn drop(&mut self) {
        let objsize = std::mem::size_of::<T>();
        let layout = Layout::from_size_align(self.size * objsize, 8).unwrap();
        unsafe {
            dealloc(self.ptr as *mut u8, layout);
        }
    }
}
