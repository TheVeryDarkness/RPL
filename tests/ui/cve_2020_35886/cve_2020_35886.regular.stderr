error: it is an undefined behavior to offset a pointer using an unchecked integer
  --> tests/ui/cve_2020_35886/cve_2020_35886.rs:36:18
   |
LL |         unsafe { self.ptr.wrapping_offset(idx as isize).as_ref() }.unwrap()
   |                  --------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                  |
   |                  offset here
   |                  pointer used here
   |
   = help: check whether it's in bound before offsetting
   = note: See the safety section in https://doc.rust-lang.org/std/primitive.pointer.html#method.offset
   = note: `-D rpl::unchecked-pointer-offset` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rpl::unchecked_pointer_offset)]`

error: it is an undefined behavior to offset a pointer using an unchecked integer
  --> tests/ui/cve_2020_35886/cve_2020_35886.rs:43:18
   |
LL |         unsafe { self.ptr.wrapping_offset(idx as isize).as_mut() }.unwrap()
   |                  --------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                  |
   |                  offset here
   |                  pointer used here
   |
   = help: check whether it's in bound before offsetting
   = note: See the safety section in https://doc.rust-lang.org/std/primitive.pointer.html#method.offset

error: aborting due to 2 previous errors

